core/git_command.py

```diff
@@ -127,8 +132,24 @@ def git(self, *args,
                                  cwd=working_dir,
                                  env=environ,
                                  startupinfo=startupinfo)
-            stdout, stderr = p.communicate(
-                (stdin.encode(encoding=stdin_encoding) if encode else stdin) if stdin else None)
+
+            if log_stderr or log_stdout:
```

Comment:
It would be cool if we could write the code  a bit like this 
```

            if stdin:
                p.stdin.write(stdin.encode(encoding=stdin_encoding) if encode else stdin)
                p.stdin.flush()
                p.stdin.close()

            if show_panel:
                util.log.panel_append(stdin + "\n\n")
                if savvy_settings.get("show_input_in_output"):
                    util.log.panel(command_str + "\n\n")

            captured_output = b''
            for line in p.stdout:
                captured_output = captured_output + line
                if show_panel:
                    util.log.panel_append(self.decode_stdout(captured_output, savvy_settings))

            process.wait()
            stdout = captured_output + process.stdout.read()
            stderr = process.stderr.read()

            stdout = self.decode_stdout(stdout, savvy_settings)
            if show_panel:
                util.log.panel_append(self.decode_stdout(stdout, savvy_settings))

            stderr = self.decode_stdout(stderr, savvy_settings)
            if show_panel and len(stderr) > 0:
                util.log.panel_append("STDERR\n\n" + self.decode_stdout(stdout, savvy_settings))
```

I did not run this code so I don't think you can copy it completely.

And a  different thing, can a git command have stdout and stderr both as streams? if so should those be in two different threads?

Reply:
From this comment I see a few suggestions:
* new logic should respect `show_input_in_output` :+1:
* new logic should replace old reliable code using `Popen.communicate()` to reduce duplication and branching
  * We'll have to _really_ test this new logic :grin: